File: packages/scheduling-engine/src/scheduler/repair-violations.ts
// =====================================================================
//  INCOMPOL PLAN — Schedule Violation Repair
//  Post-scheduling safety net: fix setup overlaps and overcapacity.
//
//  Runs after mergeConsecutiveBlocks and before deadline enforcement.
//  Catches edge cases from levelLoad/autoRouteOverflow interactions.
//
//  Pure function — no React, no side effects.
// =====================================================================

import { DAY_CAP, MINUTES_PER_DAY, S0, S2 } from '../constants.js';
import type { Block } from '../types/blocks.js';

// ── Setup Overlap Repair ─────────────────────────────────────────

/**
 * Fix setup crew overlaps by delaying later setups.
 *
 * Simulates the SetupCrew constraint on the final block set.
 * When two setups on different machines overlap in absolute time,
 * the later one is delayed to start after the earlier one ends.
 * Production start is adjusted accordingly, and production is clipped
 * if it would exceed the day boundary.
 */
function repairSetupOverlaps(blocks: Block[]): { blocks: Block[]; repaired: number } {
  const withSetup = blocks.filter(
    (b) => b.setupS != null && b.setupE != null && b.type === 'ok',
  );
  if (withSetup.length < 2) return { blocks, repaired: 0 };

  // Sort by absolute setup start time
  withSetup.sort((a, b) => {
    const absA = a.dayIdx * MINUTES_PER_DAY + a.setupS!;
    const absB = b.dayIdx * MINUTES_PER_DAY + b.setupS!;
    return absA - absB;
  });

  // Simulate setupCrew: track booked slots
  const booked: Array<{ start: number; end: number; machineId: string }> = [];
  let repaired = 0;

  for (const block of withSetup) {
    const absStart = block.dayIdx * MINUTES_PER_DAY + block.setupS!;
    const setupDur = block.setupE! - block.setupS!;

    // Find next available slot (same logic as createSetupCrew.findNextAvailable)
    let candidate = absStart;
    let changed = true;
    let iterations = 0;
    while (changed && iterations < 200) {
      changed = false;
      iterations++;
      for (const s of booked) {
        if (s.machineId === block.machineId) continue; // same machine is fine
        if (candidate < s.end && candidate + setupDur > s.start) {
          candidate = s.end;
          changed = true;
        }
      }
    }

    // Book this slot
    booked.push({ start: candidate, end: candidate + setupDur, machineId: block.machineId });

    // If we had to delay, update the block
    if (candidate !== absStart) {
      const newDay = Math.floor(candidate / MINUTES_PER_DAY);
      const newSetupS = candidate % MINUTES_PER_DAY;
      const newSetupE = newSetupS + setupDur;
      const prodDur = block.endMin - block.startMin;

      block.dayIdx = newDay;
      block.setupS = newSetupS;
      block.setupE = newSetupE;
      block.startMin = newSetupE;
      block.endMin = newSetupE + prodDur;

      // Clip if production extends past midnight (S1=1440)
      const dayEnd = 1440; // S1 — 2-shift day end
      if (block.endMin > dayEnd) {
        const clipped = block.endMin - dayEnd;
        block.endMin = dayEnd;
        block.prodMin = Math.max(0, block.prodMin - clipped);
        if (block.qty > 0 && block.prodMin > 0) {
          const origProd = prodDur;
          block.qty = Math.round(block.qty * (block.prodMin / origProd));
        }
      }

      repaired++;
    }
  }

  return { blocks, repaired };
}

// ── Overcapacity Repair ──────────────────────────────────────────

/**
 * Fix machine overcapacity by clipping the last block on overloaded days.
 *
 * For each machine-day where total minutes exceed effective capacity,
 * reduces the production of the last-ending block to fit within capacity.
 * Excess is converted to overflow.
 *
 * Accounts for overtime (overtimeMap) and 3rd shift.
 */
function repairOvercapacity(
  blocks: Block[],
  thirdShift?: boolean,
  overtimeMap?: Record<string, Record<number, number>>,
): { blocks: Block[]; repaired: number } {
  const baseCap = thirdShift ? S2 - S0 : DAY_CAP;
  let repaired = 0;

  // Group blocks by machine-day
  const mdMap = new Map<string, { machineId: string; dayIdx: number; blocks: Block[] }>();
  for (const b of blocks) {
    if (b.type !== 'ok') continue;
    const key = `${b.machineId}:${b.dayIdx}`;
    if (!mdMap.has(key)) mdMap.set(key, { machineId: b.machineId, dayIdx: b.dayIdx, blocks: [] });
    mdMap.get(key)!.blocks.push(b);
  }

  for (const [, { machineId, dayIdx, blocks: dayBlocks }] of mdMap) {
    let totalMin = 0;
    for (const b of dayBlocks) {
      totalMin += b.endMin - b.startMin;
      if (b.setupS != null && b.setupE != null) totalMin += b.setupE - b.setupS;
    }

    // Effective capacity includes overtime for this machine-day
    const ot = overtimeMap?.[machineId]?.[dayIdx] ?? 0;
    const eDayCap = baseCap + ot;

    if (Math.round(totalMin) <= eDayCap) continue;

    const excess = Math.round(totalMin) - eDayCap;

    // Find the block with the latest endMin — clip it
    dayBlocks.sort((a, b) => b.endMin - a.endMin);
    const victim = dayBlocks[0];

    const prodDur = victim.endMin - victim.startMin;
    const clipMin = Math.min(excess, prodDur);

    if (clipMin >= prodDur) {
      // Entire block becomes overflow
      victim.type = 'overflow';
    } else {
      // Clip production from the end
      const newProd = prodDur - clipMin;
      if (victim.qty > 0 && prodDur > 0) {
        victim.qty = Math.round(victim.qty * (newProd / prodDur));
      }
      victim.endMin -= clipMin;
      victim.prodMin = Math.max(0, victim.prodMin - clipMin);

      // Create overflow block for the clipped portion
      const overflowBlock: Block = {
        ...victim,
        startMin: victim.endMin,
        endMin: victim.endMin + clipMin,
        qty: Math.round((victim.qty / Math.max(1, newProd)) * clipMin),
        prodMin: clipMin,
        setupMin: 0,
        setupS: null,
        setupE: null,
        type: 'overflow',
      };
      blocks.push(overflowBlock);
    }

    repaired++;
  }

  return { blocks, repaired };
}

// ── Main Export ───────────────────────────────────────────────────

/**
 * Post-scheduling repair pass. Fixes:
 * 1. Setup crew overlaps (2+ setups on different machines simultaneously)
 * 2. Machine overcapacity (day exceeding effective DAY_CAP)
 *
 * Returns the repaired block array and a count of fixes applied.
 * When no violations exist, returns blocks unchanged.
 */
export function repairScheduleViolations(
  blocks: Block[],
  thirdShift?: boolean,
  overtimeMap?: Record<string, Record<number, number>>,
): { blocks: Block[]; setupRepairs: number; capacityRepairs: number } {
  // Work on a mutable copy
  const result = blocks.map((b) => ({ ...b }));

  const { repaired: setupRepairs } = repairSetupOverlaps(result);
  const { blocks: finalBlocks, repaired: capacityRepairs } = repairOvercapacity(
    result,
    thirdShift,
    overtimeMap,
  );

  return { blocks: finalBlocks, setupRepairs, capacityRepairs };
}

File: apps/backend/src/domain/scheduling/scheduler/repair_violations.py
"""Schedule violation repair — port of scheduler/repair-violations.ts.

Post-scheduling safety net: fix setup overlaps and overcapacity.
"""

from __future__ import annotations

from ..constants import DAY_CAP, MINUTES_PER_DAY, S0, S2
from ..types import Block


def _repair_setup_overlaps(blocks: list[Block]) -> int:
    """Fix setup crew overlaps by delaying later setups."""
    with_setup = [
        b for b in blocks if b.setup_s is not None and b.setup_e is not None and b.type == "ok"
    ]
    if len(with_setup) < 2:
        return 0

    with_setup.sort(key=lambda b: b.day_idx * MINUTES_PER_DAY + (b.setup_s or 0))

    booked: list[dict] = []
    repaired = 0

    for block in with_setup:
        abs_start = block.day_idx * MINUTES_PER_DAY + block.setup_s
        setup_dur = block.setup_e - block.setup_s

        candidate = abs_start
        changed = True
        iters = 0
        while changed and iters < 200:
            changed = False
            iters += 1
            for s in booked:
                if s["machineId"] == block.machine_id:
                    continue
                if candidate < s["end"] and candidate + setup_dur > s["start"]:
                    candidate = s["end"]
                    changed = True

        booked.append(
            {"start": candidate, "end": candidate + setup_dur, "machineId": block.machine_id}
        )

        if candidate != abs_start:
            new_day = candidate // MINUTES_PER_DAY
            new_setup_s = candidate % MINUTES_PER_DAY
            new_setup_e = new_setup_s + setup_dur
            prod_dur = block.end_min - block.start_min

            block.day_idx = new_day
            block.setup_s = new_setup_s
            block.setup_e = new_setup_e
            block.start_min = new_setup_e
            block.end_min = new_setup_e + prod_dur

            day_end = 1440
            if block.end_min > day_end:
                clipped = block.end_min - day_end
                block.end_min = day_end
                block.prod_min = max(0, block.prod_min - clipped)
                if block.qty > 0 and block.prod_min > 0:
                    block.qty = round(block.qty * (block.prod_min / prod_dur))

            repaired += 1

    return repaired


def _repair_overcapacity(
    blocks: list[Block],
    third_shift: bool = False,
    overtime_map: dict[str, dict[int, int]] | None = None,
) -> int:
    """Fix machine overcapacity by clipping the last block on overloaded days."""
    base_cap = S2 - S0 if third_shift else DAY_CAP
    repaired = 0

    md_map: dict[str, list[Block]] = {}
    for b in blocks:
        if b.type != "ok":
            continue
        key = f"{b.machine_id}:{b.day_idx}"
        if key not in md_map:
            md_map[key] = []
        md_map[key].append(b)

    for key, day_blocks in md_map.items():
        parts = key.split(":")
        machine_id, day_idx = parts[0], int(parts[1])

        total_min = 0
        for b in day_blocks:
            total_min += b.end_min - b.start_min
            if b.setup_s is not None and b.setup_e is not None:
                total_min += b.setup_e - b.setup_s

        ot = overtime_map.get(machine_id, {}).get(day_idx, 0) if overtime_map else 0
        e_day_cap = base_cap + ot

        if round(total_min) <= e_day_cap:
            continue

        excess = round(total_min) - e_day_cap
        day_blocks.sort(key=lambda b: -b.end_min)
        victim = day_blocks[0]

        prod_dur = victim.end_min - victim.start_min
        clip_min = min(excess, prod_dur)

        if clip_min >= prod_dur:
            victim.type = "overflow"
        else:
            new_prod = prod_dur - clip_min
            if victim.qty > 0 and prod_dur > 0:
                victim.qty = round(victim.qty * (new_prod / prod_dur))
            victim.end_min -= clip_min
            victim.prod_min = max(0, victim.prod_min - clip_min)

            overflow_block = victim.model_copy()
            overflow_block.start_min = victim.end_min
            overflow_block.end_min = victim.end_min + clip_min
            overflow_block.qty = (
                round((victim.qty / max(1, new_prod)) * clip_min) if victim.qty > 0 else 0
            )
            overflow_block.prod_min = clip_min
            overflow_block.setup_min = 0
            overflow_block.setup_s = None
            overflow_block.setup_e = None
            overflow_block.type = "overflow"
            blocks.append(overflow_block)

        repaired += 1

    return repaired


def repair_schedule_violations(
    blocks: list[Block],
    third_shift: bool = False,
    overtime_map: dict[str, dict[int, int]] | None = None,
) -> tuple[list[Block], int, int]:
    """Post-scheduling repair pass.

    Returns (blocks, setup_repairs, capacity_repairs).
    """
    result = [b.model_copy() for b in blocks]
    setup_repairs = _repair_setup_overlaps(result)
    capacity_repairs = _repair_overcapacity(result, third_shift, overtime_map)
    return result, setup_repairs, capacity_repairs
